Simplify memory experiment

perlinm opened the pull request:
A first implementation of my suggestion [here](https://github.com/qLDPCOrg/qLDPC/pull/257#discussion_r2249059719).  Note that docstrings have not yet been updated, and there is more code cleanup to do, but I wanted to at least package together the core changes before merging them into the `circuit_simulations` branch.

~If+when this is merged, I would like to revisit the question of whether to use [`detector_error_model_to_check_matrices`](https://github.com/quantumgizmos/ldpc/blob/8b0b47d40f1d8828aa248f8ed5df6629cd9946f7/src_python/ldpc/ckt_noise/dem_matrices.py#L61) from `ldpc` instead of writing our own implementation.~  UPDATE: I've decided to use our own implementation.

perlinm commented:
One strange observation: running the [example notebook](https://github.com/qLDPCOrg/qLDPC/blob/c4bdd6398df1299d12d849c859f73646d4c36c3c/examples/circuit_simulations.ipynb) and looking at `d=7` and `p=1e-3` with `max_shots=10**7`, I find a lower logical error rate using the check matrices constructed by @jviszlai than those constructed in `ldpc`.  I suspect that this is due to the use of different `priors`.  I'll dig further...

jviszlai commented:
This is an interesting observation! I'm torn though on if an explicit handling of such errors should be made here or if that should be handled by the decoder. In a sense they are distinct error patterns, even if they are indistinguishable when only looking at the check matrix. For example, if this is being passed off to a maximum likelihood decoder then both errors should be considered. I think this is line with your "correct" way to handle this.

I agree though that for the basically any other decoder which only looks at the check matrix this case becomes a coin flip. I'm not opposed to the idea of a solution that produces an upper bound on the logical error rate. But I'm not sure declaring the observable as flipped accomplishes that, as it's still making an assumption on the coin flip. I would think an upper bound would come from declaring any decoder prediction including such errors as a logical failure, which could maybe be handled in [`sinter_decoders.py`](https://github.com/qLDPCOrg/qLDPC/blob/circuit_simulations/qldpc/stim/sinter_decoders.py).

Also another point but I assume these two errors cannot simultaneously occur? Otherwise both occurring would correspond to a logical operator of weight 2 thus a code distance of 2.

perlinm commented:
Ah, good point re: declaring the observable as flipped is not the right thing to do.  I guess one option is to declare the observable flipped if it is more likely than not to be so, given the inferred circuit error.  That should be easy enough (albeit annoying) to compute.

I think I'm convinced though that keeping otherwise indistinguishable error mechanisms distinct and letting the decoder sort it out is the way to go.  I'll make this the default option.

Re: weight-2 logical errors, my example is indeed taken from a distance-2 code that I was playing with for debugging purposes.  So maybe this is a red herring for understanding the difference in logical error rates that I was seeing...

perlinm commented:
@jviszlai please squash+merge if+when you are happy with this PR.

Once this PR is merged into `circuit_simulations`, please open a new PR to merge `sinter_decoder.py` and `sinter_decoder_test.py` into `main`?  To modularize #257 🙂

perlinm commented:
I also cleaned up `syndrome_measurement.py` and added `syndrome_measurement_test.py` in this PR.  It may have been nice to do that in different branch/PR, but the merge conflicts for `memory.py` would get messy...

jviszlai commented on qldpc/circuits/syndrome_measurement.py:
```diff
@@ -18,135 +18,224 @@
+        Args:
+            qubit: the qubit (by index) whose measurement record we want.
+            measurement_index: an index specifying which measurement of the specified qubit we want.
+                A measurement_index of 0 would be the first measruement of the qubit, while a
```
nit typo: 'measruement -> measurement'

jviszlai replied:
Also I like this qubit-indexed way of tracking stim measurements 🙂.

jviszlai reviewed (approved):
This all seems good to me! I'm happy to merge if everything is set.
